### PR TITLE
[SYCL-MLIR][driver] Add `-fsycl-raise-host` option

### PR DESCRIPTION
**Pull request by victor-eds**
Modify driver behavior so that, if this option is found, the host code will always be compiled and raised to MLIR. Then, cgeist will receive this MLIR module as an argument via the `-sycl-use-host-module` option.

When this option is passed, two SYCL compilations are needed: one to generate the stub file to perform host compilation and another one to perform the actual device compilation (which will receive the host code as an input in MLIR format). The former one will only produce an LLVM module, so it will not be used to generate the final output.

For the device compilation to receive the host module as an input, we need to:
- Drop the regular compilation from the pipeline and add a new one which will receive both the regular input and the raised host module. This will be generated using the new `MLIRTranslate` tool.
- As the whole compilation pipeline will change if the new option is passed, the offloading should take care of it, so the host code should still be discarded if `-fsycl-device-only` is passed at this stage.
- If `-fsycl-device-only` is passed, `-sycl-device-only` will be passed to `cgeist`.
- If the host action expects an ASM output, the pipeline will be changed to avoid compiling the host twice:
  - An additional compile action with `Asm` output is inserted so that we compile the host code to LLVM bytecode and run two tools on this output: 1. `mlir-translate`, to raise to MLIR; 2. `clang` to produce the object file for the host.

2023-05-08-host-compilation-flow-3(1)

### CODE REVIEW

File: clang/lib/Driver/Driver.cpp
Comment (sommerlukas) [inline clang/lib/Driver/Driver.cpp:4877]: Nit: Can we have parentheses here to make the expression clearer?

File: clang/lib/Driver/Driver.cpp
Comment (sommerlukas) [inline clang/lib/Driver/Driver.cpp:5117]:
```suggestion
      // If the only purpose of this builder was to create the stub file, we
```

**Review by sommerlukas — COMMENTED** [review]
Can we add some documentation for this approach?

File: clang/lib/Driver/ToolChain.cpp
Comment (etiotto) [inline clang/lib/Driver/ToolChain.cpp:768]: Can you add a comment here to explain the intention of the code ?

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (etiotto) [inline clang/lib/Driver/ToolChains/Clang.cpp:7386]: If CLANG_ENABLE_OPAQUE_POINTERS_INTERNAL is false and OPT_fsycl_raise_host is true can we emit a warning stating that the new option (OPT_fsycl_raise_host) cannot be used rather than silently ignore CLANG_ENABLE_OPAQUE_POINTERS_INTERNAL

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (etiotto) [inline clang/lib/Driver/ToolChains/Clang.cpp:10179]: Why do we need to pass `-sycl-device-only=false` explicitly rather than passing nothing?

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (etiotto) [inline clang/lib/Driver/ToolChains/Clang.cpp:10220]: [note]: I guess is ok to have a loop even if currently the number of input files is expected to be one.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:2672]: We did not have a need to override this member function. Can you explain the reason we need to override it now ?

File: polygeist/tools/cgeist/driver.cc
Comment (etiotto) [inline polygeist/tools/cgeist/driver.cc:321]: Can we at least run this pass on the gpu module ? Or do you think is easier to fix the issue causing infinite recursion (in a future PR) and live with this for now ?

File: clang/test/Driver/sycl-mlir.cpp
Comment (etiotto) [inline clang/test/Driver/sycl-mlir.cpp:22]: I was expecting to see that we start a device compilation (to produce the stub). Why do we have these 3 lines ?

File: clang/test/Driver/sycl-mlir.cpp
Comment (etiotto) [inline clang/test/Driver/sycl-mlir.cpp:51]: Same comment as at line 22

File: clang/lib/Driver/Driver.cpp
Comment (etiotto) [inline clang/lib/Driver/Driver.cpp:4841]: Can you put this line back to reduce the possibility of merge conflicts?

File: clang/lib/Driver/Driver.cpp
Comment (etiotto) [inline clang/lib/Driver/Driver.cpp:5116]: Remove this line

File: clang/test/Driver/sycl-mlir.cpp
Comment (whitneywhtsang) [inline clang/test/Driver/sycl-mlir.cpp:247]: looks like we are using cgeist to generate a LLVM bitcode file, and use it as input to mlir-translate.
Shouldn't the host bitcode be generated using clang?

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (victor-eds) [reply clang/lib/Driver/ToolChains/Clang.cpp:10175]: I changed the code so that this option only works with opaque pointers.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (victor-eds) [reply clang/lib/Driver/ToolChains/Clang.cpp:10216]: Yes, that's what other tools were doing, so I went on with the trend

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (victor-eds) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Fixed already :+1:

File: polygeist/tools/cgeist/Lib/clang-mlir.cc
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: Not needed, just forgot to remove after doing some tests

File: clang/test/Driver/sycl-mlir.cpp
Comment (victor-eds) [reply clang/test/Driver/sycl-mlir.cpp:22]: That's to produce the preprocessed file that will serve as an input to the compiler together with the stub file (built in phase 5 in this example)

File: clang/test/Driver/sycl-mlir.cpp
Comment (victor-eds) [reply clang/test/Driver/sycl-mlir.cpp:51]: Same reason

File: clang/lib/Driver/Driver.cpp
Comment (victor-eds) [reply clang/lib/Driver/Driver.cpp:4841]: This was added [here](https://github.com/intel/llvm/commit/9152ef89097daf1b58b1c007e73128b0c8c101c1#diff-f8c64e36dfe3828a6f816c4550e78bb0305769ace1be53207e86ac9a3280ac9e), not in upstream DPC++. It is not needed, so I dropped that.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (victor-eds) [reply clang/lib/Driver/ToolChains/Clang.cpp:10174]: As defined [here](https://github.com/intel/llvm/commit/ad9f3b399707c8a630180d07b1f7dcc3775b7956), `cgeist` will drop the host code every time unless the output is MLIR and the option being added here is passed. We do not have to check MLIR output, only `-fsycl-device-only`.

File: clang/test/Driver/sycl-mlir.cpp
Comment (victor-eds) [reply clang/test/Driver/sycl-mlir.cpp:247]: That is the stub file being built. Do you want me to add a check for `clang` in between to make this more explicit?

File: polygeist/tools/cgeist/driver.cc
Comment (etiotto) [reply polygeist/tools/cgeist/driver.cc:321]: Ok

**Review by etiotto — APPROVED** [review]

File: clang/test/Driver/sycl-mlir.cpp
Comment (whitneywhtsang) [reply clang/test/Driver/sycl-mlir.cpp:247]: Yes, I think that would be useful.

**Comment by victor-eds** [comment]
We should wait for @Naghasan approval here, as he's the most knowledgeable on the driver in our side. I'll mark this as draft to avoid this being merged.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (Naghasan) [inline clang/lib/Driver/ToolChains/Clang.cpp:10164]:
```suggestion
    if (I.getType() == types::TY_MLIR_IR && I.Act->isHostOffloading(Action::OFK_SYCL))
```

Nitpick, As this should map to the LLVM IR -> MLIR compile action, that should capture this is actually the host (and not a device MLIR)

**Review by Naghasan — APPROVED** [review]
A few nitpick, I think there is a better way than duplicating the SYCLActionBuilder but it's for a hack anyway so will revisit that only if we need to.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**
